Add cast bingtile to/from bigint

jagill opened the pull request:
Externally, tiles are encoded in a string of chars '0' to '3' called a
quadkey.  Internally, Presto encodes a tile in 64 bits, represented by a
BIGINT.  Storing a tile as a bigint is not only more space/cpu efficient
than storing it as a quadkey, but it also avoids the bucket-skew problem
caused by the non-uniform distribution of `hash(quadkey) mod 2^k`.

```
== RELEASE NOTES ==

General Changes
* Add `cast(tile AS bigint)` and `cast(bigint_value AS bingtile)` to encode/decode Bing tiles to/from bigints.  This is a more efficient storage format that also reduces bucket skew in some cases.
```

jagill commented:
@mbasmanova

jagill commented:
One question I have is whether we should think about versioning the encoding, now that it may be used to persist tiles.  We have 13 bits left to play with...

jagill commented:
cc @tdcmeehan

tdcmeehan commented:
The purpose of these functions seems to be casting from the BingTile type to/from BIGINT.  If that's the case, why not just adding operators to cast to/from BIGINT instead of these functions?

tdcmeehan commented on presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestBingTileFunctions.java:
```diff
@@ -80,6 +82,32 @@ public void testSerialization()
+    @Test
+    public void testBingTileBigint()
+    {
+        Random random = new Random(1812);
```
I would do this in a loop of say 1000 just to make sure we get good signal.

jagill replied:
Since this is testing a query, the setup/teardown meant that doing 1000 (particularly for each level) was really slow (a couple minutes on my laptop).  Is there a way to reduce the overhead, or is that kind of test length worth the coverage?

tdcmeehan replied:
Instead of 1000, any nontrivial positive number would probably work

jagill commented:
@mbasmanova and I chatted about whether to view this as a cast, or to use an explicit function.  Our thinking was that since this is not a canonical transformation, but rather a encoding/decoding, it made sense to make the function explicit.  What are your thoughts on that?

tdcmeehan commented:
We do similar things already.  For example, internally HLL is represented as `Slice`, and we support cast to `VARBINARY`--which is also an encoding/decoding.  To me this seems like the same sort of thing, except the internal data type is a `long` rather than a `Slice`.  So if we accept this premise, then I think cast is the way to go for consistency (both in terms of how it works for other types, and also it being a consistent convention in Presto).

jagill commented:
@tdcmeehan Ok, I replaced the functions with cast operations, and made the tests iterate 20 random tiles at each level (fewer for levels 0, 1, 2 when there are less than 20 tiles).

jagill commented:
Did you have any thoughts on versioning the tile serialization?

tdcmeehan commented:
> Did you have any thoughts on versioning the tile serialization?

I think it makes sense to reserve some of the bits for versioning.  It's understandable that someone would want to leverage this feature to save storage space as well.

tdcmeehan commented on presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/BingTileFunctions.java:
```diff
@@ -85,6 +87,29 @@
+            BingTile.decode(tile);
+        }
+        catch (IllegalArgumentException e) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT,
```
```suggestion
            throw new PrestoException(INVALID_CAST_ARGUMENT,
```

jagill commented:
@tdcmeehan I've addressed your comments by:
1. Making the highest 5 bits a version, moving the zoom to the highest-of-32 bits.  As a nice side benefit, this allows us to extend to zoom 27 by just changing a constant.
2. Changing the error to INVALID_CAST_ARGUMENT
3. Making an encoding unit test, which tests up to 1000 samples for each zoom level,
4. Reducing the functional test to up to 5 samples per zoom level.

arhimondr commented on presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/BingTileFunctions.java:
```diff
@@ -85,6 +88,29 @@
+    public static long castFromBigint(@SqlType(StandardTypes.BIGINT) long tile)
+    {
+        try {
+            BingTile.decode(tile);
```
follow up: Currently `BingTile.decode` must create an object of `BingTile`. The main reason for calling `BingTile.decode` is to validate the tile stored in a long. For efficiency I would recommend to have a dedicated method, e.g.: `BingTile.validate` or smthng.

arhimondr commented on presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestBingTileFunctions.java:
```diff
@@ -80,6 +82,56 @@ public void testSerialization()
     }
 
+    @Test
+    void testBingTileEncoding()
```
nit: public

arhimondr commented on presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestBingTileFunctions.java:
```diff
@@ -80,6 +82,56 @@ public void testSerialization()
+        for (int zoom = 0; zoom <= MAX_ZOOM_LEVEL; zoom++) {
+            int numIterations = Math.min(1000, (int) Math.pow(4, zoom));
+            for (int iter = 0; iter < numIterations; iter++) {
+                BingTile expected = BingTile.fromCoordinates(
```
nit: extract these 3 lines into something like `testRoundTrip(BingTile tile)`

arhimondr commented on presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestBingTileFunctions.java:
```diff
@@ -80,6 +82,56 @@ public void testSerialization()
+    @Test
+    void testBingTileEncoding()
+    {
+        Random random = new Random(1491);
```
I'm not sure if this code really has to be fuzzed.

How about standard test cases:

- Min `zoom`
- Max `zoom`
- Min `x` and `y`
- Max `x` and `y`
- Several different combinations of values in between

arhimondr commented on presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestBingTileFunctions.java:
```diff
@@ -80,6 +82,56 @@ public void testSerialization()
+    public void testBingTileCast()
+    {
+        Random random = new Random(1812);
+        for (int zoom = 0; zoom <= MAX_ZOOM_LEVEL; zoom++) {
```
This test is only supposed to verify the integration of the tile encoding (that is tested in the test above), and the function mechanism. Instead of fuzzing i would recommend adding just a few simple test cases to verify the integration is in place.

jagill commented:
@arhimondr I've address your comments, modulo setting up the separate validation function.  As per our conversation, doing that work require either annoying duplicated code, or additional heap allocations.  So we'll deal with that if there are performance issues associated with the current method.
